src/main/java/de/tum/in/www1/artemis/service/GradingScaleService.java

```diff
@@ -0,0 +1,118 @@
+        }
+        gradingScale.setGradeSteps(gradeSteps);
+        return gradingScaleRepository.save(gradingScale);
+    }
+
+    /**
+     * Checks the validity of a grade step set and throws an exception if one of the following conditions is not fulfilled
+     * - all individuals grade steps should be in a valid format
+     * - the grade steps set should form a valid and congruent grading scale
+     *
+     * @param gradeSteps the grade steps to be checked
+     */
+    private void checkGradeStepValidity(Set<GradeStep> gradeSteps) {
+        if (gradeSteps != null) {
+            if (!gradeSteps.stream().allMatch(GradeStep::isValid)) {
+                throw new BadRequestAlertException("Not all grade steps are following the correct format.", "gradeStep", "invalidFormat");
+            }
+            else if (!gradeStepSetMapsToValidGradingScale(gradeSteps)) {
+                throw new BadRequestAlertException("Grade step set can't match to a valid grading scale.", "gradeStep", "invalidFormat");
+            }
+        }
+    }
+
+    /**
+     * Checks if the grade steps map to a valid grading scale
+     * - the grade names should all be unique for the grading scale
+     * - when ordered, all steps should fulfill valid adjacency
+     * - the first and the last element should fulfill the boundary conditions (start with 0% and end with 100%)
+     *
+     * @param gradeSteps the grade steps to be checked
+     * @return true if the grade steps map to a valid grading scale and false otherwise
+     */
+    private boolean gradeStepSetMapsToValidGradingScale(Set<GradeStep> gradeSteps) {
+        if (gradeSteps.stream().map(GradeStep::getGradeName).distinct().count() != gradeSteps.size()) {
+            return false;
+        }
+        List<GradeStep> sortedGradeSteps = gradeSteps.stream().sorted(Comparator.comparingDouble(GradeStep::getLowerBoundPercentage)).collect(Collectors.toList());
+        boolean validAdjacency = IntStream.range(0, sortedGradeSteps.size() - 1).allMatch(i -> GradeStep.checkValidAdjacency(sortedGradeSteps.get(i), sortedGradeSteps.get(i + 1)));
+        boolean validFirstElement = sortedGradeSteps.get(0).isLowerBoundInclusive() && sortedGradeSteps.get(0).getLowerBoundPercentage() == 0;
+        boolean validLastElement = sortedGradeSteps.get(sortedGradeSteps.size() - 1).isUpperBoundInclusive()
+                && sortedGradeSteps.get(sortedGradeSteps.size() - 1).getUpperBoundPercentage() == 100;
+        return validAdjacency && validFirstElement && validLastElement;
+    }
+
+    /**
+     * Deletes a grading scale and all its grade steps
+     *
+     * @param gradingScale the grading scale which will be deleted
+     */
+    public void delete(GradingScale gradingScale) {
+        List<GradeStep> gradeSteps = gradeStepRepository.findByGradingScale_Id(gradingScale.getId());
+        gradeStepRepository.deleteInBatch(gradeSteps);
+        gradingScaleRepository.deleteInBatch(List.of(gradingScale));
```

Comment:
I fail to see why a batch delete is needed here? Maybe I am thinking wrong but it should be enough to just delete the gradingScale as the delete will be propagated to all the steps anyway

Reply:
Tbh I tried a lot of different ways to delete the grading scale, and this turned out the only one I've managed to find that works for sure. I've tried using a simple delete, which should in theory delete the grading scale and propagate to the grade steps because of the relationship definition, but for some reason hibernate doesn't execute the delete statements at all (I've turned on the additional logging locally), and neither the scale nor the steps get deleted.